README.md
<a href="http://lua.org">
<img src="http://vinniefalco.github.com/LuaBridgeDemo/powered-by-lua.png">
</a><br>

# LuaBridge 2.5

[LuaBridge][1] is a lightweight and dependency-free library for mapping data,
functions, and classes back and forth between C++ and [Lua][2] (a powerful,
fast, lightweight, embeddable scripting language) . LuaBridge has been tested
and works with Lua revisions starting from 5.1.5., although it should work in
any version of Lua from 5.1.0 as well as [LuaJit][3].

LuaBridge offers the following features:

- [MIT Licensed][4]
- A printable [Reference Manual][5].
- Headers-only: No Makefile, no .cpp files, just one #include!
- Simple, light, and nothing else needed (like Boost).
- No macros, settings, or configuration scripts needed.
- Supports different object lifetime management models.
- Convenient, type-safe access to the Lua stack.
- Automatic function parameter type binding.
- Easy access to Lua objects like tables and functions.
- Written in a clear and easy to debug style.
- Does not require C++11.

Please read the [LuaBridge Reference Manual][5] for more details on the API.

## Unit Tests

Unit test build requires a CMake and C++11 compliant compiler.

There are 4 unit test flavors:
* `Tests51` - uses Lua 5.1, with C++11 features
* `Tests51L` - uses Lua 5.1, no C++11 features
* `Tests52` - uses Lua 5.2, with C++11 features
* `Tests52L` - uses Lua 5.2, no C++11 features

Build using Make on Linux/MacOS:
```bash
clone --recurse-submodules git@github.com:vinniefalco/LuaBridge.git
cd LuaBridge
cmake -DCMAKE_BUILD_TYPE=Debug -B build
# or cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -B build
# or cmake -DCMAKE_BUILD_TYPE=Release -B build
cd build
make -j
```

Generate XCode project on MacOS:
```bash
clone --recurse-submodules git@github.com:vinniefalco/LuaBridge.git
cd LuaBridge
cmake -G Xcode -B build
# Generates XCode project build/LuaBridge.xcodeproj
```

Generate MSVS solution on Windows:
```cmd
clone --recurse-submodules git@github.com:vinniefalco/LuaBridge.git
cd LuaBridge
mkdir build
cd build
cmake -G "Visual Studio 15 2017 Win64" -B build
# or cmake -G "Visual Studio 14 2015" -B build
# or cmake -G "Visual Studio 15 2017 Win64" -B build
# or cmake -G "Visual Studio 15 2017" -B build
# or cmake -G "Visual Studio 15 2019" -A Win64 -B build
# or cmake -G "Visual Studio 15 2019" -B build
# Generates MSVS solution build/LuaBridge.sln
```

# LuaBridge Demo

LuaBridge provides both a command line program and a stand-alone graphical
program for compiling and running the test suite. The graphical program brings
up an interactive window where you can enter execute Lua statements in a
persistent environment. This application is cross platform and works on
Windows, Mac OS, iOS, Android, and GNU/Linux systems with X11. The stand-alone
program should work anywhere. Both of these applications include LuaBridge,
Lua version 5.2, and the code necessary to produce a cross platform graphic
application. They are all together in a separate repository, with no
additional dependencies, available on Github at [LuaBridge Demo and Tests][6].
This is what the GUI application looks like, along with the C++ code snippet
for registering the two classes:

<a href="https://github.com/vinniefalco/LuaBridgeDemo">
<img src="http://vinniefalco.github.com/LuaBridgeDemo/LuaBridgeDemoScreenshot1.0.2.png">
</a><br>

## Official Repository

LuaBridge is published under the terms of the [MIT License][4].

The original version of LuaBridge was written by Nathan Reed. The project has
been taken over by [Vinnie Falco][7], who added new functionality, wrote the new
documentation, and incorporated contributions from Nigel Atkinson.

For questions, comments, or bug reports feel free to open a Github issue
or contact Vinnie Falco directly at the email address indicated below.

Copyright 2019, Dmitry Tarakanov<br>
Copyright 2012, [Vinnie Falco][7] (<[vinnie.falco@gmail.com][8]>)<br>
Copyright 2008, Nigel Atkinson<br>
Copyright 2007, Nathan Reed<br>

Portions from The Loki Library:<br>
Copyright (C) 2001 by Andrei Alexandrescu

Older versions of LuaBridge up to and including 0.2 are distributed under the
BSD 3-Clause License. See the corresponding license file in those versions
(distributed separately) for more details.

[1]:  https://github.com/vinniefalco/LuaBridge "LuaBridge"
[2]:  http://lua.org "The Lua Programming Language"
[3]:  http://luajit.org/ "The LuaJIT Probject"
[4]:  http://www.opensource.org/licenses/mit-license.html "The MIT License"
[5]:  http://vinniefalco.github.com/LuaBridge "LuaBridge Reference Manual"
[6]:  https://github.com/vinniefalco/LuaBridgeDemo "LuaBridge Demo"
[7]:  https://github.com/vinniefalco "Vinnie Falco's Github"
[8]:  mailto:vinnie.falco@gmail.com "Vinnie Falco (Email)"


CHANGES.txt
2.0.1:
Jeffrey Carpenter - Sat Nov 29 12:06:05 2014
 * Fixed image colorspace issue on iOS and Mac OS X
Sam Lantinga - Sun Jun 15 17:33:46 2014
 * Fixed support for transparency in XPM files
Davide Coppola - Thu Apr 17 17:30:12 2014
 * Fixed building JPEG support on Android
David Ludwig - Tue Apr 01 19:40:35 2014
 * Added support for building for Windows RT and Windows Phone
Timur - Wed Dec 11 21:24:36 2013
 * Fixed memory leak in webp image loading
Patrice Mandin - Thu Nov 07 19:15:28 2013
 * Fixed loading BMP files with large BITMAPINFOHEADER structures
Sam Lantinga - Fri Oct 11 21:54:20 2013
 * Fixed building with libpng 1.4

2.0.0:
Sam Lantinga - Sun Jun  2 22:25:31 PDT 2013
 * Added PNG save support based on miniz.c by Rich Geldreich
       IMG_SavePNG(), IMG_SavePNG_RW()
Sam Lantinga - Sat Jun  1 19:11:26 PDT 2013
 * Updated for SDL 2.0 release
Sam Lantinga - Sat Mar 23 13:36:51 PDT 2013
 * Fixed bug setting colorkey for indexed PNG images
Torsten Stremlau - Sun Mar 10 10:19:25 PDT 2013
 * Added support for alpha and lossless WEBP images

1.2.12:
mscott - 2012-02-06 19:40:23 PST
 * Fixed image corruption when using ImageIO framework
Sylvain - Thu Nov 22 13:09:59 PST 2012
 * Added extended XPM color table (disabled by default in IMG_xpm.c)
Sam Lantinga - Thu Jan 19 23:18:09 EST 2012
 * Fixed regression in 1.2.11 loading 8-bit PNG images with libpng

1.2.11:
Sam Lantinga - Sat Jan 14 17:54:38 EST 2012
 * Fixed loading 8-bit PNG images on Mac OS X
Sam Lantinga - Sat Dec 31 09:35:40 EST 2011
 * SDL_image is now under the zlib license
Michael Bonfils - Mon Nov 28 21:46:00 EST 2011
 * Added WEBP image support
Thomas Klausner - Wed Jan 19 19:31:25 PST 2011
 * Fixed compiling with libpng 1.4
Sam Lantinga - Mon Jan 10 12:09:57 2011 -0800
 * Added Android.mk to build on the Android platform
Sam Lantinga - Mon May 10 22:42:53 PDT 2010
 * Fixed loading HAM6 images with stencil mask
Mark Tucker - Fri, 27 Nov 2009 12:38:21 -0500
 * Fixed bug loading 15 and 16 bit BMP images

1.2.10:
Sam Lantinga - Sat Nov 14 11:22:14 PST 2009
 * Fixed bug loading multiple images

1.2.9:
Sam Lantinga - Tue Nov 10 00:29:20 PST 2009
 * Fixed alpha premultiplication on Mac OS X and iPhone OS
Sam Lantinga - Sun Nov  8 07:52:11 PST 2009
 * Fixed checking for IMG_Init() return value in image loaders

1.2.8:
Sam Lantinga - Sun Oct  4 13:12:54 PDT 2009
 * Added support for uncompressed PCX files
Mason Wheeler - 2009-06-10 06:29:45 PDT 
 * Added IMG_Init()/IMG_Quit() to prevent constantly loading and unloading DLLs
Couriersud - Mon, 12 Jan 2009 17:21:13 -0800
 * Added support for ICO and CUR image files
Eric Wing - Fri, 2 Jan 2009 02:01:16 -0800
 * Added ImageIO loading infrastructure for Mac OS X
 * Added UIImage loading infrastructure for iPhone / iPod Touch

1.2.7:
Sam Lantinga - Sun Nov  2 15:08:27 PST 2008
 * Fixed buffer overflow in BMP loading code, discovered by j00ru//vx
Sam Lantinga - Fri Dec 28 08:34:54 PST 2007
 * Fixed buffer overflow in GIF loading code, discovered by Michael Skladnikiewicz

1.2.6:
Sam lantinga - Wed Jul 18 00:30:32 PDT 2007
 * Improved detection of libjpeg, libpng, and libtiff at configure time
 * PNG and TIFF images are correctly identified even if dynamic libraries
   to load them aren't available.
 * Fixed loading of TIFF images using libtiff 3.6
Sam Lantinga - Thu Jul 5 07:52:35 2007
 * Fixed static linking with libjpeg
Michael Koch - Tue Feb 13 10:09:17 2007
 * Fixed crash in IMG_ReadXPMFromArray()

1.2.5:
Maurizio Monge - Sun May 14 13:57:32 PDT 2006
 * Fixed loading BMP palettes at unusual offsets
Sam Lantinga - Thu May 11 21:51:19 PDT 2006
 * Added support for dynamically loading libjpeg, libpng, and libtiff.
Sam Lantinga - Sun Apr 30 01:48:40 PDT 2006
 * Added gcc-fat.sh for generating Universal binaries on Mac OS X
 * Updated libtool support to version 1.5.22
Sam Lantinga - Sat Feb  4 15:17:44 PST 2006
 * Added support for XV thumbnail images
Gautier Portet - Fri, 19 Mar 2004 17:35:12 +0100
 * Added support for 32-bit BMP files with alpha

1.2.4:
Pierre G. Richard - Fri, 30 Jul 2004 11:13:11 +0000 (UTC)
 * Added support for RLE encoded BMP files
Marc Le Douarain - Fri, 26 Dec 2003 18:23:42 +0100
 * Added EHB and HAM mode support to the ILBM loader
Sam Lantinga - Wed Nov 19 00:23:44 PST 2003
 * Updated libtool support for new mingw32 DLL build process
Holger Schemel - Mon, 04 Aug 2003 21:50:52 +0200
 * Fixed crash loading certain PCX images
Kyle Davenport - Sat, 19 Apr 2003 17:13:31 -0500
 * Added .la files to the development RPM, fixing RPM build on RedHat 8

1.2.3:
Ryan C. Gordon - Sat, 8 Feb 2003 09:36:33 -0500
 * Fixed memory leak with non-seekable SDL_RWops
Marc Le Douarain - Sun, 22 Dec 2002 22:59:51 +0100
 * Added 24-bit support to the ILBM format loader
Sam Lantinga - Sun Oct 20 20:55:46 PDT 2002
 * Added shared library support for MacOS X
Pete Shinners - Thu Jun 20 10:05:54 PDT 2002
 * The JPEG loader can now load EXIF format JPEG images
Dag-Erling Smorgrav - Thu May  2 19:09:48 PDT 2002
 * The XCF loader now ignores invisible layers and channels

1.2.2:
Sam Lantinga - Sat Apr 13 07:49:47 PDT 2002
 * Updated autogen.sh for new versions of automake
 * Specify the SDL API calling convention (C by default)


celfile.cpp
#include <stdio.h>
#include <stdint.h>

#include <SDL/SDL.h>
#include <SDL/SDL_image.h>
#include <iostream>
#include <vector>
#include <algorithm>


#include "celfile.h"
#include "cel_frame.h"


void get_pal(std::string pal_filename, colour* pal)
{
    FILE * pal_file;

    pal_file = fopen(pal_filename.c_str(), "rb");
    
    for(int i = 0; i < 256; i++)
    {
            fread(&pal[i].r, 1, 1, pal_file);
            fread(&pal[i].g, 1, 1, pal_file);
            fread(&pal[i].b, 1, 1, pal_file);
    }

    fclose(pal_file);
}

size_t Cel_file::read_num_frames()
{
    fseek(mFile, 0, SEEK_SET);

    uint32_t num_frames;
    
    fread(&num_frames, 4, 1, mFile);

    #ifdef CEL_DEBUG
        std::cout << ftell(mFile) << ": Num frames: " << num_frames << std::endl;
    #endif

    return num_frames;
}

void Cel_file::read_frame_offsets()
{
    fseek(mFile, 4, SEEK_SET);

    for(size_t i = 0; i < mNum_frames; i++){
            fread(&mFrame_offsets[i], 4, 1, mFile);
            //std::cout << ftell(mFile) << ": offset " << i << ": " << mFrame_offsets[i] << std::endl;
    }

    fread(&mFrame_offsets[mNum_frames], 4, 1, mFile);

    #ifdef CEL_DEBUG
        std::cout << ftell(mFile) << ": end offset: " << mFrame_offsets[mNum_frames] << std::endl;
    #endif
}

void Cel_file::fix_image(std::vector<colour>& raw_image, size_t width)
{
    for(int i = 0; i < raw_image.size()/2; i++)
    {
        colour tmp = raw_image[i];

        int indextemp = raw_image.size();
        indextemp -= i;

        //if(indextemp > raw_image.size() -1)
        //    std::cout << "---------------------------------" << std::endl;

        colour tmp2 = raw_image[indextemp];
        raw_image[i] = tmp2;
        raw_image[raw_image.size() - i] = tmp;
    }

    for(int i = 0; i < raw_image.size(); i++)
    {
        int x = i%width;
        if(x < width/2)
        {
            colour tmp = raw_image[i];
            raw_image[i] = raw_image[(i-x+width) - x];
            raw_image[(i-x+width) - x] = tmp;
        }
    }
}

void fill_t(size_t pixels, std::vector<colour>& raw_image)
{
    for(int px = 0; px < pixels; px++)
        raw_image.push_back(colour(255, 255, 255, false));
}


bool Cel_file::greater_than_first(uint8_t* frame, size_t frame_size)
{
    return frame_size >= 196 &&
    frame[  2] == 0 && frame[  3] == 0 &&
    frame[ 14] == 0 && frame[ 15] == 0 &&
    frame[ 34] == 0 && frame[ 35] == 0 &&
    frame[ 62] == 0 && frame[ 63] == 0 &&
    frame[ 98] == 0 && frame[ 99] == 0 &&
    frame[142] == 0 && frame[143] == 0 &&
    frame[194] == 0 && frame[195] == 0;
}

bool Cel_file::greater_than_second(uint8_t* frame, size_t frame_size)
{
    return frame_size >= 196 &&
    frame[254] == 0 && frame[255] == 0 &&
    frame[318] == 0 && frame[319] == 0 &&
    frame[374] == 0 && frame[375] == 0 &&
    frame[422] == 0 && frame[423] == 0 &&
    frame[462] == 0 && frame[463] == 0 &&
    frame[494] == 0 && frame[495] == 0 &&
    frame[518] == 0 && frame[519] == 0 &&
    frame[534] == 0 && frame[535] == 0;
}

bool Cel_file::is_greater_than(uint8_t* frame, size_t frame_size)
{
    return greater_than_first(frame, frame_size);
}

void Cel_file::decode_greater_than(uint8_t* frame, size_t frame_size, colour* pal, std::vector<colour>& raw_image)
{
    #ifdef CEL_DEBUG
        std::cout << "Greater than" << std::endl;
    #endif

    int segment;
    int i = 0;



    raw_image.push_back(pal[frame[i]]);
    i++;
    raw_image.push_back(pal[frame[i]]);
    i++;
    fill_t(30, raw_image); // finish off the first line
   

    for(segment = 0; segment < 7; segment++)
    {
        i += 2;

        int xdraw = (segment+1)*4;
        int xoffs = 0;
        
        #ifdef CEL_DEBUG
            std::cout << "\tdraw: " << xdraw << std::endl;
        #endif

        int px; 
        for(px = xoffs; px < xdraw; px++)
        {
                raw_image.push_back(pal[frame[i]]);
                i++;
        }
        fill_t(32-xdraw, raw_image);
        

        
        xdraw = (segment+1)*4 +2;
        xoffs = 0;

        #ifdef CEL_DEBUG
            std::cout << "\tdraw: " << xdraw << std::endl;
        #endif
        
        for(px = xoffs; px < xdraw; px++)
        {
                raw_image.push_back(pal[frame[i]]);
                i++;
        }
        fill_t(32-xdraw, raw_image);
        
        
       
        #ifdef CEL_DEBUG
            std::cout << "len: " << raw_image.size() << std::endl;
        #endif
    }
    
    if(greater_than_second(frame, frame_size))
    {
        for(; segment < 15; segment++)
        {
            i += 2;

            int xdraw = (15-segment)*4;
            int xoffs = 0;

            #ifdef CEL_DEBUG
                std::cout << "\tdraw: " << xdraw << std::endl;
            #endif

            int px; 
            for(px = xoffs; px < xdraw; px++)
            {
                    raw_image.push_back(pal[frame[i]]);
                    i++;
            }
            fill_t(32-xdraw, raw_image);
            

            
            xdraw = ((15-segment)*4) -2;
            xoffs = 0;

            #ifdef CEL_DEBUG
                std::cout << "\tdraw: " << xdraw << std::endl;
            #endif

            
            for(px = xoffs; px < xdraw; px++)
            {
                    raw_image.push_back(pal[frame[i]]);
                    i++;
            }
            fill_t(32-xdraw, raw_image);
            
            
            #ifdef CEL_DEBUG 
                std::cout << "len: " << raw_image.size() << std::endl;
            #endif
        }

        fill_t(32, raw_image); // last transparent line
    }
    else
    {
        for(int i = 256; i < frame_size; i++)
            raw_image.push_back(pal[frame[i]]);
    }
    
    #ifdef CEL_DEBUG
        std::cout << "GT" << raw_image.size() << std::endl;
    #endif

}

bool less_than_first(uint8_t* frame, size_t frame_size)
{
    return frame_size >= 226 &&
    frame[  0] == 0 && frame[  1] == 0 &&
    frame[  8] == 0 && frame[  9] == 0 &&
    frame[ 24] == 0 && frame[ 25] == 0 &&
    frame[ 48] == 0 && frame[ 49] == 0 &&
    frame[ 80] == 0 && frame[ 81] == 0 &&
    frame[120] == 0 && frame[121] == 0 &&
    frame[168] == 0 && frame[169] == 0 &&
    frame[224] == 0 && frame[225] == 0;
}

bool less_than_second(uint8_t* frame, size_t frame_size)
{
    return frame_size >= 530 &&
    frame[288] == 0 && frame[289] == 0 &&
    frame[348] == 0 && frame[349] == 0 &&
    frame[400] == 0 && frame[401] == 0 &&
    frame[444] == 0 && frame[445] == 0 &&
    frame[480] == 0 && frame[481] == 0 &&
    frame[508] == 0 && frame[509] == 0 &&
    frame[528] == 0 && frame[529] == 0;
}

bool is_less_than(uint8_t* frame, size_t frame_size)
{
    return less_than_first(frame, frame_size);
}





void decode_less_than(uint8_t* frame, size_t frame_size, colour* pal, std::vector<colour>& raw_image)
{
    #ifdef CEL_DEBUG
        std::cout << "Less Than" << std::endl;
    #endif

    int segment;
    int i = 0;

    for(segment = 0; segment < 8; segment++)
    {
        i += 2;

        int xdraw = segment*4 +2;
        int xoffs = 32 - xdraw;
        
        #ifdef CEL_DEBUG
            std::cout << "\tdraw: " << xdraw << std::endl;
            std::cout << "\toff: " << xoffs << std::endl;
        #endif

           
        fill_t(xoffs, raw_image);
        for(int px = xoffs; px < 32; px++)
        {
                raw_image.push_back(pal[frame[i]]);
                i++;
        }
        

        
        xdraw = segment*4 +4;
        xoffs = 32 - xdraw;

        #ifdef CEL_DEBUG
            std::cout << "\tdraw: " << xdraw << std::endl;
        #endif

        
        fill_t(xoffs, raw_image);
        for(int px = xoffs; px < 32; px++)
        {
                raw_image.push_back(pal[frame[i]]);
                i++;
        }
        
        #ifdef CEL_DEBUG
            std::cout << "len: " << raw_image.size() << std::endl;
        #endif
    }
    
    if(less_than_second(frame, frame_size))
    {
        for(; segment < 16; segment++)
        {
            i += 2;

            int xdraw = (15-segment)*4 +2;
            int xoffs = 32 - xdraw;

            #ifdef CEL_DEBUG
                std::cout << "\tdraw2: " << xdraw << std::endl;
            #endif

            fill_t(xoffs, raw_image);
               
            for(int px = xoffs; px < 32; px++)
            {
                    raw_image.push_back(pal[frame[i]]);
                    i++;
            }
            

            
            xdraw = (15-segment)*4;
            xoffs = 32 - xdraw;
        
            #ifdef CEL_DEBUG
                std::cout << "\tdraw2: " << xdraw << std::endl;
            #endif

            
            fill_t(xoffs, raw_image);
            for(int px = xoffs; px < 32; px++)
            {
                    raw_image.push_back(pal[frame[i]]);
                    i++;
            }
            
            #ifdef CEL_DEBUG
                std::cout << "len2: " << raw_image.size() << std::endl;
            #endif
        }
    }
    else
    {
        for(int i = 256; i < frame_size; i++)
            raw_image.push_back(pal[frame[i]]);
    }

}

int32_t normal_width(uint8_t* frame, size_t frame_size, bool from_header, uint16_t offset)
{
    
    // If we have a header, we know that offset points to the start of the 32nd line.
    // So, when we reach that point, we will have produced 31 lines of pixels, so we 
    // can divide the number of pixels we have passed at this point by 31, to get the 
    // width.
    if(from_header)
    {
        int32_t width_header = 0; 
        
        for(size_t i = 11; i < frame_size; i++){
            
            if(i == offset && from_header)
            {
                width_header = width_header/31;
                break;
            }
            // Regular command
            if(frame[i] <= 127){
                width_header += frame[i];
                i += frame[i];
            }

            // Transparency command
            else if(128 <= frame[i]){
                width_header += 256 - frame[i];
            }
        }

        return width_header;
    }
    
    // If we do not have a header we probably (definitely?) don't have any transparency.
    // The maximum stretch of opaque pixels following a command byte is 127.
    // Since commands can't wrap over lines (it seems), if the width is shorter than 127,
    // the first (command) byte will indicate an entire line, so it's value is the width.
    // If the width is larger than 127, it will be some sequence of 127 byte long stretches,
    // followed by some other value to bring it to the end of a line (presuming the width is
    // not divisible by 127).
    // So, for all image except those whose width is divisible by 127, we can determine width
    // by looping through control bits, adding 127 each time, until we find some value which
    // is not 127, then add that to the 127-total and that is our width.
    else
    {
        int32_t width_reg = 0;
        
        for(size_t i = 0; i < frame_size; i++){

            // Regular command
            if(frame[i] <= 127){
                width_reg += frame[i];
                i += frame[i];
            }

            // Transparency command - who knows, it might be possible
            else if(128 <= frame[i]){
                width_reg += 256 - frame[i];
            }

            if(frame[i] != 127)
                break;

        }

        return width_reg;
    }
}


void normal_decode(uint8_t* frame, size_t frame_size, size_t width, bool from_header, colour* pal, std::vector<colour>& raw_image)
{
    #ifdef CEL_DEBUG
        std::cout << "NORMAL_DECODE" << std::endl;
    #endif

    size_t i;

    // Skip the header if it exists
    if(from_header)
        i = 11;
    else
        i = 0;
    
    for(; i < frame_size; i++)
    {
        // Regular command
        if(frame[i] <= 127)
        {
            #ifdef CEL_DEBUG
                std::cout << i << " regular: " << (int)frame[i] << std::endl;
            #endif
            
            size_t j;
            // Just push the number of pixels specified by the command
            for(j = 1; j < frame[i]+1 && i+j < frame_size; j++)
            {
                int index = i+j;
                uint8_t f = frame[index];
                
                if(index > frame_size-1)
                    std::cout << "invalid read from f " << index << " " << frame_size << std::endl;

                colour col = pal[f];

                raw_image.push_back(col);
            }
            //if(i+j >= frame_size)
            //    std::cout << "----- " << i << " " << j << " " << i+j << " " << frame_size << std::endl;
            
            i+= frame[i];
        }

        // Transparency command
        else if(128 <= frame[i])
        {
            #ifdef CEL_DEBUG
                std::cout << i << " transparency: " << (int)frame[i] << " " << (256 - frame[i]) << std::endl;
            #endif
            
            // Push (256 - command value) transparent pixels
            for(size_t j = 0; j < 256-frame[i]; j++)
                raw_image.push_back(colour(255, 0, 255, false));
        }
    }
}

size_t decode_raw_32(uint8_t* frame, size_t frame_size, colour* pal, std::vector<colour>& raw_image)
{

    for(int i = 0; i < frame_size; i++)
    {
        raw_image.push_back(pal[frame[i]]);
    }

    return 32;
}

Cel_frame& Cel_file::operator[] (size_t index)
{
    if(mCache.count(index))
        return mCache[index];


    Cel_frame frame;
    std::vector<colour> raw_image;
    frame.raw_image = raw_image;
    frame.width = get_frame(index, frame.raw_image);
    frame.height = frame.raw_image.size() / frame.width;

    mCache[index] = frame;
    
    #ifdef CEL_DEBUG
        std::cout << "w: " << frame.width << ", h: " << frame.height << std::endl;
    #endif

    return mCache[index];
}

size_t Cel_file::get_frame(size_t frame_num, std::vector<colour>& raw_image)
{
    size_t frame_size = mFrame_offsets[frame_num+1] - mFrame_offsets[frame_num];

    #ifdef CEL_DEBUG
        std::cout << std::endl << "frame 0 size: " << frame_size << std::endl;
    #endif

    

    // Load frame data
    uint8_t frame[frame_size];
    fseek(mFile, mFrame_offsets[frame_num], SEEK_SET);
    fread(frame, 1, frame_size, mFile);

    // Make sure we're not concatenating onto some other image 
    raw_image.clear();
    

    //print_cel(frame, frame_size);
    //std::cout << std::endl;
   
    size_t width;
    
    if(mIs_tile_cel)
        width = 32;

    if(mIs_tile_cel && is_less_than(frame, frame_size))
        decode_less_than(frame, frame_size, mPal, raw_image);
    
    else if(mIs_tile_cel && is_greater_than(frame, frame_size))
        decode_greater_than(frame, frame_size, mPal, raw_image);
    
    else
    {
        uint16_t offset;
        bool from_header = false;
        
        // Tile cel frames never have headers 
        if(!mIs_tile_cel)
        { 
            // The frame has a header which we can use to determine width
            if(frame[0] == 10)
            {
                from_header = true;
                fseek(mFile, mFrame_offsets[frame_num]+2, SEEK_SET);
                fread(&offset, 2, 1, mFile);
            }
            
            width = normal_width(frame, frame_size, from_header, offset);
        }
        
        normal_decode(frame, frame_size, width, from_header, mPal, raw_image);
        
        #ifdef CEL_DEBUG
            std::cout << raw_image.size() << " " << frame_size << std::endl;
        #endif
        
        if(mIs_tile_cel && frame_size == 1024 && frame_num != 2593) // It's a fully opaque raw frame, width 32, from a level tileset
        //if(mIs_tile_cel && raw_image.size() < frame_size ) // It's a fully opaque raw frame, width 32, from a level tileset
        {
            raw_image.clear();
            decode_raw_32(frame, frame_size, mPal, raw_image);
        }
        

    }
    
    //fix_image(raw_image, width);
    
    #ifdef CEL_DEBUG 
        std::cout << "WIDTH used: " << width << std::endl;
    #endif


    return width;
}

bool ends_with(const std::string& full, const std::string& end)
{
    return end.size() <= full.size() && full.substr(full.size() - end.size(), end.size()) == end;
}

std::string replace_end(const std::string& old_end, const std::string& new_end, const std::string& original)
{
    std::string retval = original.substr(0, original.size() - old_end.size());
    retval.append(new_end);
    return retval;
}

bool is_tile_cel(const std::string& file_name)
{
    return 
    ends_with(file_name, "l1.cel") ||
    ends_with(file_name, "l2.cel") ||
    ends_with(file_name, "l3.cel") ||
    ends_with(file_name, "l4.cel") ||
    ends_with(file_name, "town.cel");
}

size_t Cel_file::get_num_frames()
{
    return mNum_frames;
}

void get_pallette(std::string filename, colour* pal)
{
    std::string pal_filename;
    if(ends_with(filename, "l1.cel"))
        pal_filename = replace_end("l1.cel", "l1.pal", filename);
    else
        pal_filename = "diablo.pal";

    
    return get_pal(pal_filename, pal);
}

Cel_file::Cel_file(std::string filename)
{
    mFile = fopen(filename.c_str(), "rb");

    mPal = new colour[256];
    //get_pal("diablo.pal", mPal);
    get_pallette(filename, mPal);

    mNum_frames = read_num_frames();

    mFrame_offsets = new uint32_t[mNum_frames+1]; 
    //get_frame_offsets(mFile, mFrame_offsets, mNum_frames);
    read_frame_offsets();
    
    mIs_tile_cel = is_tile_cel(filename);
}


